aggregator/list.go

```diff
@@ -85,88 +128,81 @@ func newMetricListMetrics(scope tally.Scope) metricListMetrics {
 
 type flushBeforeFn func(beforeNanos int64, flushType flushType)
 
-// metricList stores aggregated metrics at a given resolution
-// and flushes aggregations periodically.
-type metricList struct {
+// baseMetricList is a metric list storing aggregations at a given resolution and
+// flushing them periodically.
+type baseMetricList struct {
 	sync.RWMutex
 
-	shard         uint32
-	opts          Options
-	nowFn         clock.NowFn
-	log           log.Logger
-	timeLock      *sync.RWMutex
-	flushHandler  handler.Handler
-	flushWriter   writer.Writer
-	resolution    time.Duration
-	flushInterval time.Duration
-	flushMgr      FlushManager
-
-	closed                   bool
-	aggregations             *list.List
-	lastFlushedNanos         int64
-	toCollect                []*list.Element
+	shard            uint32
+	opts             Options
+	nowFn            clock.NowFn
+	timeLock         *sync.RWMutex
+	flushHandler     handler.Handler
+	localWriter      writer.Writer
+	forwardedWriter  client.AdminClient
+	resolution       time.Duration
+	isEarlierThanFn  isEarlierThanFn
+	timestampNanosFn timestampNanosFn
+
+	closed           bool
+	aggregations     *list.List
+	lastFlushedNanos int64
+	toCollect        []*list.Element
+	metrics          baseMetricListMetrics
+
 	flushBeforeFn            flushBeforeFn
 	consumeLocalMetricFn     flushLocalMetricFn
 	discardLocalMetricFn     flushLocalMetricFn
 	consumeForwardedMetricFn flushForwardedMetricFn
 	discardForwardedMetricFn flushForwardedMetricFn
-	metrics                  metricListMetrics
 }
 
-func newMetricList(shard uint32, resolution time.Duration, opts Options) (*metricList, error) {
+func newBaseMetricList(
+	shard uint32,
+	resolution time.Duration,
+	isEarlierThanFn isEarlierThanFn,
+	timestampNanosFn timestampNanosFn,
+	opts Options,
+) (*baseMetricList, error) {
 	scope := opts.InstrumentOptions().MetricsScope().SubScope("list").Tagged(
 		map[string]string{"resolution": resolution.String()},
 	)
 	flushHandler := opts.FlushHandler()
-	flushWriter, err := flushHandler.NewWriter(scope.SubScope("writer"))
+	localWriter, err := flushHandler.NewWriter(scope.SubScope("writer"))
 	if err != nil {
 		return nil, err
 	}
 
-	// NB(xichen): by default the flush interval is the same as metric
-	// resolution, unless the resolution is smaller than the minimum flush
-	// interval, in which case we use the min flush interval to avoid excessing
-	// CPU overhead due to flushing.
-	flushInterval := resolution
-	if minFlushInterval := opts.MinFlushInterval(); flushInterval < minFlushInterval {
```

Comment:
you don't need this MinFlushInterval logic any more?

Reply:
Yeah I gave this more thought and realized that when the resolution is high (1s), chances are that users do want to see those metrics get flushed more quickly, so I got rid of the restriction here. We should probably add capacity or use a dedicated agg tier for those super high res metrics instead of using a longer flush interval since that kinda defeats the purpose of having high res metrics.

Additionally, this makes it difficult to reason what the max tolerated lateness should be given the metric resolution and number of forwarded times, since the resolution is not always the same as the flush interval.